libs/flexIPC/fleXdUDS/src/FleXdUDSServer.cpp

```diff
@@ -0,0 +1,141 @@
+#include "FleXdIPCMsg.h"
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <cstring>
+#include <map>
+
+namespace flexd {
+    namespace icl {
+        namespace ipc {
+
+            FleXdUDSServer::FleXdUDSServer(const std::string& socPath, FleXdEpoll& poller, FleXdIPC* proxy)
+            : FleXdUDS(socPath, poller, proxy) {
+            }
+
+            FleXdUDSServer::~FleXdUDSServer() {
+            }
+
+            bool FleXdUDSServer::initUDS() {
+                if (listenUDS()) {
+                    m_poller.addEvent(getFd(), [this](FleXdEpoll::Event evn) {
+                        m_proxy->rcvEvent(evn);
+                    });
+                    return true;
+                }
+                return false;
+            }
+
+            void FleXdUDSServer::rcvEvent(FleXdEpoll::Event e) {
+                if (e.type == EpollEvent::EpollIn) {
+                    // self fd is listen fd for incoming connections -> accept will perform
+                    if(e.fd == getFd()) {
+                        int clientFd = ::accept(getFd(), NULL, NULL);
+                        auto it = m_map.find(clientFd);
+                        if (it == m_map.end()) {
+                            m_proxy->connectClient(clientFd);
+                        }
+                    // other fd's are client fd's -> read will perform
+                    } else {
+                        int rc = 1;
+                        byteArray8192 array;
+                        while ((rc = ::read(e.fd, &array[0], sizeof (array))) > 0) {
+                            readMsg(e, std::move(array), rc);
+                        }
+                    }
+                } else if (e.type == EpollEvent::EpollError) {
+                    if(e.fd != getFd()) {
+                        m_proxy->disconnectClient(e.fd);
+                    }
+                }
+            }
+
+            void FleXdUDSServer::connectClient(int fd) {
+                ::fcntl(fd, F_SETFL, O_NONBLOCK);
+                FleXdIPCBuffer buffer(fd, [this](pSharedFleXdIPCMsg msg, int fd) {
+                    m_proxy->rcvMsg(msg, fd);
+                });
+                m_map.insert(std::make_pair(fd, std::move(buffer)));
+                m_poller.addEvent(fd, [this](FleXdEpoll::Event evn) {
+                    m_proxy->rcvEvent(evn);
+                });
+            }
+
+            void FleXdUDSServer::disconnectClient(int fd) {
+                std::ignore = m_map.erase(fd);
+            }
+
+            void FleXdUDSServer::readMsg(FleXdEpoll::Event e, std::array<uint8_t, 8192>&& array, int size) {
+                std::shared_ptr<std::array<uint8_t, 8192> > array_ptr = std::make_shared<std::array<uint8_t, 8192> >(array);
+                auto search = m_map.find(e.fd);
+                if (search != m_map.end()) {
+                    search->second.rcvMsg(array_ptr, size);
```

Comment:
Constructing 'shared_ptr' before could be considered as mispractice since there's an evaluation which may not always pass. Therefore, calling 'make_shared' to construct 'byteArray8192' inside 'rcvMsg' is much better since there will be no additional overhead and no resources would have to be freed if evaluation fails.


TL;DR: Don't allocate resources before calling 'if' - if not necessary. And use 'byteArray8192'